sdk/core/core-websockets/src/retry.ts

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { calculateRetryDelay, delay } from "@azure/core-util";
+import { logger } from "./logger.js";
+import { createError } from "./utils.js";
+
+export const Constants = {
+  defaultOperationTimeoutInMs: 60000,
+  reconnectLimit: 100,
+  aadTokenValidityMarginInMs: 5000,
+  connectionReconnectDelay: 300,
+  defaultMaxRetries: 3,
+  defaultMaxRetriesForConnection: 150,
+  defaultDelayBetweenOperationRetriesInMs: 30000,
+  defaultMaxDelayForExponentialRetryInMs: 90000,
+} as const;
+
+/**
+ * Describes the Retry Mode type
+ */
+export type RetryMode = "Exponential" | "Fixed";
+
+/**
+ * Retry policy options that determine the mode, number of retries, retry interval etc.
+ */
+export interface RetryOptions {
+  /**
+   * Number of times the operation needs to be retried in case
+   * of retryable error. Default: 3.
+   */
+  maxRetries?: number;
+  /**
+   * Amount of time to wait in milliseconds before making the
+   * next attempt. Default: `30000 milliseconds`.
+   * When `mode` option is set to `Exponential`,
+   * this is used to compute the exponentially increasing delays between retries.
+   */
+  retryDelayInMs?: number;
+  /**
+   * Number of milliseconds to wait before declaring that current attempt has timed out which will trigger a retry
+   * A minimum value of `60000` milliseconds will be used if a value not greater than this is provided.
+   */
+  timeoutInMs?: number;
+  /**
+   * Denotes which retry mode to apply. If undefined, defaults to `Exponential`
+   */
+  mode?: RetryMode;
+  /**
+   * Denotes the maximum delay between retries
+   * that the retry attempts will be capped at. Applicable only when performing exponential retry.
+   */
+  maxRetryDelayInMs?: number;
+}
+
+/**
+ * Calculates delay between retries, in milliseconds.
+ */
+function calculateDelay(
+  attemptCount: number,
+  retryDelayInMs: number,
+  maxRetryDelayInMs: number,
+  mode: RetryMode,
```

Comment:
Should we also think about a jitter here as well instead of just the base retry delays?